diff --git a/absl/base/config.h b/absl/base/config.h
index 63b9642d..6d09003e 100644
--- a/absl/base/config.h
+++ b/absl/base/config.h
@@ -282,7 +282,7 @@ static_assert(ABSL_INTERNAL_INLINE_NAMESPACE_STR[0] != 'h' ||
 // supported.
 #ifdef ABSL_HAVE_THREAD_LOCAL
 #error ABSL_HAVE_THREAD_LOCAL cannot be directly set
-#else
+#elif !defined(__MVS__)
 #define ABSL_HAVE_THREAD_LOCAL 1
 #endif
 
diff --git a/absl/base/internal/sysinfo.cc b/absl/base/internal/sysinfo.cc
index 1937db30..64076681 100644
--- a/absl/base/internal/sysinfo.cc
+++ b/absl/base/internal/sysinfo.cc
@@ -475,6 +475,13 @@ pid_t GetTID() {
   return static_cast<pid_t>(zx_thread_self());
 }
 
+#elif defined(__MVS__)
+
+pid_t GetTID() {
+  return static_cast<pid_t>(pthread_self().__);
+}
+
+
 #else
 
 // Fallback implementation of `GetTID` using `pthread_self`.
